Lib/fontTools/pens/ftPen.py

```diff
@@ -0,0 +1,159 @@
+from freetype.ft_errors import FT_Exception
+
+from fontTools.pens.basePen import BasePen
+from fontTools.misc.roundTools import otRound
+
+class FTPen(BasePen):
+
+    np = None
+    plt = None
+    Image = None
+    Contour = collections.namedtuple('Contour', ('points', 'tags'))
+    LINE      = 0b00000001
+    CURVE     = 0b00000011
+    OFFCURVE  = 0b00000010
+    QCURVE    = 0b00000001
+    QOFFCURVE = 0b00000000
+
+    def __init__(self, glyphSet):
+        self.contours = []
+
+    def outline(self, offset=None, scale=None, even_odd=False):
+        # Convert the current contours to FT_Outline.
+        offset = offset or (0, 0)
+        scale  = scale  or (1.0, 1.0)
+        n_contours = len(self.contours)
+        n_points   = sum((len(contour.points) for contour in self.contours))
+        points = []
+        for contour in self.contours:
+            for point in contour.points:
+                points.append(FT_Vector(FT_Pos(otRound((point[0] + offset[0]) * scale[0] * 64)), FT_Pos(otRound((point[1] + offset[1]) * scale[1] * 64))))
+        tags = []
+        for contour in self.contours:
+            for tag in contour.tags:
+                tags.append(tag)
+        contours = []
+        contours_sum = 0
+        for contour in self.contours:
+            contours_sum += len(contour.points)
+            contours.append(contours_sum - 1)
+        flags = FT_OUTLINE_EVEN_ODD_FILL if even_odd else FT_OUTLINE_NONE
+        return FT_Outline(
+            (ctypes.c_short)(n_contours),
+            (ctypes.c_short)(n_points),
+            (FT_Vector      * n_points)(*points),
+            (ctypes.c_ubyte * n_points)(*tags),
+            (ctypes.c_short * n_contours)(*contours),
+            (ctypes.c_int)(flags)
+        )
+
+    def buffer(self, width=1000, ascender=880, descender=-120, even_odd=False, scale=None):
+        # Return a tuple with the bitmap buffer and its dimension.
+        scale  = scale or (1.0, 1.0)
+        width  = math.ceil(width * scale[0])
+        height = math.ceil((ascender - descender) * scale[1])
+        buf = ctypes.create_string_buffer(width * height)
+        bitmap = FT_Bitmap(
+            (ctypes.c_int)(height),
+            (ctypes.c_int)(width),
+            (ctypes.c_int)(width),
+            (ctypes.POINTER(ctypes.c_ubyte))(buf),
+            (ctypes.c_short)(256),
+            (ctypes.c_ubyte)(FT_PIXEL_MODE_GRAY),
+            (ctypes.c_char)(0),
+            (ctypes.c_void_p)(None)
+        )
+        outline = self.outline(offset=(0, -descender), even_odd=even_odd, scale=scale)
+        err = FT_Outline_Get_Bitmap(freetype.get_handle(), ctypes.byref(outline), ctypes.byref(bitmap))
+        if err != 0:
+            raise FT_Exception(err)
+        return buf.raw, (width, height)
+
+    def array(self, width=1000, ascender=880, descender=-120, even_odd=False, scale=None):
+        # Return a numpy array. Each element takes values in the range of [0.0, 1.0].
+        if not self.np:
```

Comment:
I would not cache PIL, numpy or matplotlib: imports are cached by Python already (via `sys.modules`), so importing again doesn't add a huge cost.

Reply:
Absolutely right. Thank you for pointing this out!